File: composite/java/ProductCatalog.java
package Composite;


import java.util.ArrayList;
import java.util.List;

// Somut yapımızdır.
// UML diyagramındaki Composite kısmına denk gelmektedir.
// Component (ICatalogComponent) listesi tutmaktadır.
public class ProductCatalog implements ICatalogComponent {
    private String _name;

    private List<ICatalogComponent> _components;

    public ProductCatalog(String name) {
        _name = name;
        _components = new ArrayList<ICatalogComponent>();
    }

    public void add(ICatalogComponent catalogComponent) {
        _components.add(catalogComponent);
    }

    public void remove(ICatalogComponent catalogComponent) {
        _components.remove(catalogComponent);
    }

    public void drawHierarchy() {
        System.out.println(_name);
        for (ICatalogComponent component : _components) {
            component.drawHierarchy();
        }
    }
}

File: composite/kotlin/ProductCatalog.kt
package Composite

import java.util.ArrayList

// Somut yapımızdır.
// UML diyagramındaki Composite kısmına denk gelmektedir.
// Component (ICatalogComponent) listesi tutmaktadır.
class ProductCatalog(private val _name: String) : ICatalogComponent {

    private val _components: ArrayList<ICatalogComponent> = ArrayList()

    fun add(catalogComponent: ICatalogComponent) {
        _components.add(catalogComponent)
    }

    fun remove(catalogComponent: ICatalogComponent) {
        _components.remove(catalogComponent)
    }

    override fun drawHierarchy() {
        println(_name)
        for (component in _components) {
            component.drawHierarchy()
        }
    }
}